### PR TITLE
MAINT: move ufunc override code to umath and multiarray as much as possible.

### PR DESCRIPTION
**Pull request by mhvk**
Follow-up on #11338, with a move of all functions from `src/private` that is useful only to `umath` or `multiarray` to the files where they are actually used. No changes except to function names. Follow-up will be to get `out` only once in `override.c` (and then percolate this up).

I think this should be uncontroversial except I'm not 100% sure on best practice for the names of the functions - ideally, they give a hint about the include file they came from. (Also, eventually `GetOutObjects` should likely be moved somewhere else. But I think that can be done later.)

@eric-wieser - requesting your review since you already did #11338.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Sorry - super slow on this, and now there's a conflict

**Review by eric-wieser — APPROVED** [review]
Needs a rebase, but otherwise looks pretty reasonable - certainly the moves to `methods.c` make sense to me, although I don't have much of a feel of the distinction between the other two files.

File: numpy/core/src/private/ufunc_override.c
Comment (eric-wieser) [inline numpy/core/src/private/ufunc_override.c]: out of scope for this PR: Can't you just use `PyArray_Type` here?

**Comment by mattip** [comment]
I tried to rebase this, but one of the moved functions, `has_non_default_array_ufunc`, is now also used in `number.c` (after #11450).

**Comment by mhvk** [comment]
I think it will be easiest to rebase this after umath and multiarray are merged...

File: numpy/core/src/private/ufunc_override.c
Comment (mhvk) [reply numpy/core/src/private/ufunc_override.c]: To use `PyArray_Type` needs the umath/multiarray merge...

**Comment by mattip** [comment]
rebased and reworked. 

There is a TODO in `array_ufunc` since the function duplicates what is subsequently done in `PyUFunc_CheckOverride`. `array_ufunc` is python-level `ndarray.__array_ufunc__`. It executes the checks, then calls `PyObject_Call(ufunc_method, ...)` which eventually calls `PyUFunc_CheckOverride`. However `array_ufunc` returns `NotImplemented` when it fails to find an appropriate `ufunc`, `PyUFunc_CheckOverride` raises a `TypeError`. The cleanest way to fix this would be let `PyUFunc_CheckOverride` do the work, but to pass information between them. So either `PyUFunc_CheckOverride` needs to know it is being called from `array_ufunc` (perhaps via the `kwd` dict?) or `array_ufunc` needs to know if `PyUFunc_CheckOverride` raised a `TypeError('operand type(s) all returned NotImplemented ...')`. Perhaps there are other ways to communicate the information.

The test that shows the redunancy is in `test_umath.py`, `TestSpecialMethods.test_ufunc_override_with_super`

**Comment by mhvk** [comment]
@mattip - thanks also on this one for rebasing and reworking.  I'll try to make this one a bit more of a priority to look at, since it was part of a larger plan to try to streamline the whole override checking. Do ping if you don't hear back soon!

File: numpy/core/src/common/ufunc_override.c
Comment (eric-wieser) [inline numpy/core/src/common/ufunc_override.c]: @mattip: If you're changing this, you can now use `(PyObject *)&PyArray_Type` above too, right?

Comment (mhvk) [reply numpy/core/src/common/ufunc_override.c]: This made sense to me, but a trial gave an immediate segmentation fault...

File: numpy/core/src/common/ufunc_override.c
Comment (mhvk) [inline numpy/core/src/common/ufunc_override.c]: As noted in the main comments, one can for some reason *not* replace this with `PyArray_CheckExact` - `PyArray_Type` does not seem to be defined here. I'll check if this is just missing an include...

Comment (mhvk) [reply numpy/core/src/common/ufunc_override.c]: OK, think I found it - need `define _MULTIARRAYMODULE`.

**Comment by mhvk** [comment]
OK, now looked at this movement PR again, and thought all looked good except that I cleaned up a few of the comments. As noted, I was confused by why the `PyArray_Type` thing doesn't work, but then saw the `NO_IMPORT_ARRAY` and realized that was the problem.

File: numpy/core/src/common/ufunc_override.c
Comment (mhvk) [reply numpy/core/src/common/ufunc_override.c]: Solved with `#define _MULTIARRAYMODULE` above

File: numpy/core/src/common/ufunc_override.c
Comment (eric-wieser) [inline numpy/core/src/common/ufunc_override.c]: `import` is not really accurate any more

Comment (mhvk) [reply numpy/core/src/common/ufunc_override.c]: Can fix this; was there anything else? (Hoping to run CI just one more time...)

Comment (mattip) [reply numpy/core/src/common/ufunc_override.c]: ping

**Review by eric-wieser — APPROVED** [review]
I think this looks good to go - I don't want to spend any more time looking a this big but mostly move-only diff. I wish git would be more helpful here.

File: numpy/core/src/common/ufunc_override.c
Comment (charris) [reply numpy/core/src/common/ufunc_override.c]: This is resolved?

File: numpy/core/src/private/ufunc_override.c
Comment (charris) [reply numpy/core/src/private/ufunc_override.c]: This is resolved? Github really doesn't help here, can the committer hit the resolve button? Should they?

**Comment by charris** [comment]
Thanks Marten.

**Comment by mhvk** [comment]
Thanks, all the problems here were indeed resolved.

**Pull request merged**
